bst.py

```diff
@@ -207,6 +235,9 @@ def delete(self, val):
         self.set.remove(val)
         return None
 
+    def re_balance(self):
+        return self.root.re_balance()
+
```

Comment:
For a tree to be self-balancing, you should be calling this from inside the delete and insert methods of the tree.  This is a manually re-balanced tree.